docs/design/specs/Ecma-335-Augments.md

```diff
@@ -325,3 +325,19 @@ We propose the limit on #String and #Blob heap size is 2^29 (0.5 GB), that is an
 2) 0.5 GB is a very large heap. Having such a big PE file seems unreasonable and very rare scenario (if it exists at all).
 
 3) Having 3 spare bits available is very beneficial for the implementation. It allows to represent WinRT projected strings, namespaces, etc. in very efficient way. If we had to represent heap indices with all 32 bits it would bloat various structures and increase memory pressure. PE files over 0.5 GB of size are very rare, but the overhead would affect all compilers and tools working with the metadata reader.
+
+## Module Initializer
+
+All modules may have a module initializer. This method shall be static, a member of the module, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.
+
+There are no limitations on what code is permitted in a module initializer. Module initializers are permitted to run and call both managed and unmanaged code.
+
+### Module Initialization Guarantees
+
+The CLI shall provide the following guarantees regarding module initialization:
+
+1. A module initializer is executed at, or sometime before, first access to any static field or first invocation of any method defined in the module.
+
+2. A module initializer shall run exactly once for any given module unless explicitly called by user code.
+
+3. No method other than those called directly or indirectly from the module initializer will be able to access the types, methods, or data in a module before its initializer completes execution.
```

Comment:
Do we need to describe what happens when there are races and deadlocks between module initializers à la II.10.5.3.3 of the ECMA-335 spec?

Also would an initializer that throws an exception be considered as having "completed execution"?

Reply:
It should just say that it is same as for type initializers. Something like:

The module initializers should be treated as if it were type initializers of type `<Module>` for the purpose of guarantees about races and deadlocks (II.10.5.3.3).

Reply:
@jkotas Do I understand correctly that module initializers *are* type initializers of the `<Module>` type?

Reply:
Or:

> As the type initializer for the `<Module>` type, a module initializer has the same race and deadlock guarantees (II.10.5.3.3) as any other type initializer.

Reply:
Module initializers behave a lot like initializers of the `<Module>` type. 

If we were to say that module initializer is the type initializer of the <Module> type, I think it should be earlier in the chapter and the module initializer guarantees should be described as additional guarantees on top of the type initializer guarantees.

Reply:
Does this definition allow for the possibility that a module initializer is not a type initializer of the \<Module> type?

> This method shall be a global method, static, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.

Reply:
> Module initializers behave a lot like initializers of the `<Module>` type.
> 
> If we were to say that module initializer is the type initializer of the type, I think it should be earlier in the chapter and the module initializer guarantees should be described as additional guarantees on top of the type initializer guarantees.

@jkotas Can you confirm which of the two directions you'd like to go? I think there could be subtleties that I wouldn't be able to appreciate yet.

Reply:
Describing the module initializer guarantees as additional guarantees on top of the type initializer guarantees should make the spec simpler, so I would go with that.

Reply:
Updated. What do you think?